Data-dir env var not set properly for remote vcs dependencies

Using a remote vcs dependency (`pkg-b`) this unexpectedly fails:

```
$ git clone  https://github.com/hvr/cabal-bugs
$ cd cabal-bugs/issue-5469

$ cabal new-build all
Cloning into '/tmp/cabal-bugs/issue-5469/dist-newstyle/src/cabal-bugs-7607c4ae10cdb862'...
remote: Counting objects: 62, done.
remote: Compressing objects: 100% (12/12), done.
remote: Total 62 (delta 2), reused 16 (delta 1), pack-reused 44
Unpacking objects: 100% (62/62), done.
Checking connectivity... done.
Note: checking out 'b23f3277270b2066163fabde41a08be1261bf9eb'.
HEAD is now at b23f327... Add repro-case for haskell/cabal#5469
Resolving dependencies...
Build profile: -w ghc-8.2.2 -O1
In order, the following will be built (use -v for more details):
 - pkg-a-0 (lib) (first run)
 - pkg-b-0 (exe:b) (first run)
Configuring library for pkg-a-0..
Preprocessing library for pkg-a-0..
Building library for pkg-a-0..
[1 of 2] Compiling Paths_pkg_a      ( /tmp/cabal-bugs/issue-5469/dist-newstyle/build/x86_64-linux/ghc-8.2.2/pkg-a-0/build/autogen/Paths_pkg_a.hs, /tmp/cabal-bugs/issue-5469/dist-newstyle/build/x86_64-linux/ghc-8.2.2/pkg-a-0/build/Paths_pkg_a.o )
[2 of 2] Compiling A                ( src/A.hs, /tmp/cabal-bugs/issue-5469/dist-newstyle/build/x86_64-linux/ghc-8.2.2/pkg-a-0/build/A.o )
Configuring executable 'b' for pkg-b-0..
Preprocessing executable 'b' for pkg-b-0..
Building executable 'b' for pkg-b-0..
[1 of 1] Compiling Main             ( Main.hs, /tmp/cabal-bugs/issue-5469/dist-newstyle/build/x86_64-linux/ghc-8.2.2/pkg-b-0/x/b/build/b/b-tmp/Main.o )
filename = "/home/hvr/.cabal/share/x86_64-linux-ghc-8.2.2/pkg-a-0/datafile.txt"

Main.hs:8:9: error:
    • Exception when trying to run compile-time code:
        /home/hvr/.cabal/share/x86_64-linux-ghc-8.2.2/pkg-a-0/datafile.txt: openFile: does not exist (No such file or directory)
      Code: embedDataFileTxt
    • In the untyped splice: $(embedDataFileTxt)
  |
8 |     s = $(embedDataFileTxt)
  |         ^^^^^^^^^^^^^^^^^^^
```

However, for a simple folder-based local package (using the `cabal.ok.project` file) this works just fine:

```
$ cabal --proj=cabal.ok.project new-build all
Resolving dependencies...
Build profile: -w ghc-8.2.2 -O1
In order, the following will be built (use -v for more details):
 - pkg-a-0 (lib) (first run)
 - pkg-b-0 (exe:b) (first run)
Configuring library for pkg-a-0..
Preprocessing library for pkg-a-0..
Building library for pkg-a-0..
[1 of 2] Compiling Paths_pkg_a      ( /tmp/cabal-bugs/issue-5469/dist-newstyle/build/x86_64-linux/ghc-8.2.2/pkg-a-0/build/autogen/Paths_pkg_a.hs, /tmp/cabal-bugs/issue-5469/dist-newstyle/build/x86_64-linux/ghc-8.2.2/pkg-a-0/build/Paths_pkg_a.o )
[2 of 2] Compiling A                ( src/A.hs, /tmp/cabal-bugs/issue-5469/dist-newstyle/build/x86_64-linux/ghc-8.2.2/pkg-a-0/build/A.o )
Configuring executable 'b' for pkg-b-0..
Preprocessing executable 'b' for pkg-b-0..
Building executable 'b' for pkg-b-0..
[1 of 1] Compiling Main             ( Main.hs, /tmp/cabal-bugs/issue-5469/dist-newstyle/build/x86_64-linux/ghc-8.2.2/pkg-b-0/x/b/build/b/b-tmp/Main.o )
filename = "/tmp/cabal-bugs/issue-5469/pkg-a/datadir/datafile.txt"
Linking /tmp/cabal-bugs/issue-5469/dist-newstyle/build/x86_64-linux/ghc-8.2.2/pkg-b-0/x/b/build/b/b ...
```

So the problem is that in the VCS codepaths, the `pkg_a_datadir` env-var isn't set, wheres in the 2nd case the `pkg_a_datadir` env-var is set properly.


(cc @edsko: jfyi, this is the very bug we're experiencing w/ the cardano-sl repo)

--- a/cabal-install/Distribution/Client/CmdRun.hs
+++ b/cabal-install/Distribution/Client/CmdRun.hs
@@ -286,7 +286,9 @@ runAction (configFlags, configExFlags, installFlags, haddockFlags)
       emptyProgramInvocation {
         progInvokePath  = exePath,
         progInvokeArgs  = args,
-        progInvokeEnv   = dataDirsEnvironmentForPlan elaboratedPlan
+        progInvokeEnv   = dataDirsEnvironmentForPlan
+                            (distDirLayout baseCtx)
+                            elaboratedPlan
       }
     
     handleDoesNotExist () (removeDirectoryRecursive tempDir)

--- a/cabal-install/Distribution/Client/ProjectBuilding.hs
+++ b/cabal-install/Distribution/Client/ProjectBuilding.hs
@@ -910,7 +910,7 @@ buildAndInstallUnpackedPackage :: Verbosity
                                -> FilePath -> FilePath
                                -> IO BuildResult
 buildAndInstallUnpackedPackage verbosity
-                               DistDirLayout{distTempDirectory}
+                               distDirLayout@DistDirLayout{distTempDirectory}
                                storeDirLayout@StoreDirLayout {
                                  storePackageDBStack
                                }
@@ -1086,7 +1086,8 @@ buildAndInstallUnpackedPackage verbosity
     copyFlags destdir _ = setupHsCopyFlags pkg pkgshared verbosity
                                            builddir destdir
 
-    scriptOptions = setupHsScriptOptions rpkg plan pkgshared srcdir builddir
+    scriptOptions = setupHsScriptOptions rpkg plan pkgshared
+                                         distDirLayout srcdir builddir
                                          isParallelBuild cacheLock
 
     setup :: CommandUI flags -> (Version -> flags) -> IO ()
@@ -1099,7 +1100,7 @@ buildAndInstallUnpackedPackage verbosity
           verbosity
           scriptOptions
             { useLoggingHandle     = mLogFileHandle
-            , useExtraEnvOverrides = dataDirsEnvironmentForPlan plan }
+            , useExtraEnvOverrides = dataDirsEnvironmentForPlan distDirLayout plan }
           (Just (elabPkgDescription pkg))
           cmd flags args
 
@@ -1359,7 +1360,7 @@ buildInplaceUnpackedPackage verbosity
                        setupHsHaddockArgs pkg
 
     scriptOptions    = setupHsScriptOptions rpkg plan pkgshared
-                                            srcdir builddir
+                                            distDirLayout srcdir builddir
                                             isParallelBuild cacheLock
 
     setupInteractive :: CommandUI flags

--- a/cabal-install/Distribution/Client/ProjectPlanning.hs
+++ b/cabal-install/Distribution/Client/ProjectPlanning.hs
@@ -3108,6 +3108,7 @@ legacyCustomSetupPkgs compiler (Platform _ os) =
 setupHsScriptOptions :: ElaboratedReadyPackage
                      -> ElaboratedInstallPlan
                      -> ElaboratedSharedConfig
+                     -> DistDirLayout
                      -> FilePath
                      -> FilePath
                      -> Bool
@@ -3116,7 +3117,7 @@ setupHsScriptOptions :: ElaboratedReadyPackage
 -- TODO: Fix this so custom is a separate component.  Custom can ALWAYS
 -- be a separate component!!!
 setupHsScriptOptions (ReadyPackage elab@ElaboratedConfiguredPackage{..})
-                     plan ElaboratedSharedConfig{..} srcdir builddir
+                     plan ElaboratedSharedConfig{..} distdir srcdir builddir
                      isParallelBuild cacheLock =
     SetupScriptOptions {
       useCabalVersion          = thisVersion elabSetupScriptCliVersion,
@@ -3135,7 +3136,7 @@ setupHsScriptOptions (ReadyPackage elab@ElaboratedConfiguredPackage{..})
       useLoggingHandle         = Nothing, -- this gets set later
       useWorkingDir            = Just srcdir,
       useExtraPathEnv          = elabExeDependencyPaths elab,
-      useExtraEnvOverrides     = dataDirsEnvironmentForPlan plan,
+      useExtraEnvOverrides     = dataDirsEnvironmentForPlan distdir plan,
       useWin32CleanHack        = False,   --TODO: [required eventually]
       forceExternalSetupMethod = isParallelBuild,
       setupCacheLock           = Just cacheLock,

--- a/cabal-install/Distribution/Client/ProjectPlanning/Types.hs
+++ b/cabal-install/Distribution/Client/ProjectPlanning/Types.hs
@@ -378,12 +378,13 @@ elabRequiresRegistration elab =
 -- | Construct the environment needed for the data files to work.
 -- This consists of a separate @*_datadir@ variable for each
 -- inplace package in the plan.
-dataDirsEnvironmentForPlan :: ElaboratedInstallPlan
+dataDirsEnvironmentForPlan :: DistDirLayout
+                           -> ElaboratedInstallPlan
                            -> [(String, Maybe FilePath)]
-dataDirsEnvironmentForPlan = catMaybes
+dataDirsEnvironmentForPlan distDirLayout = catMaybes
                            . fmap (InstallPlan.foldPlanPackage
                                (const Nothing)
-                               dataDirEnvVarForPackage)
+                               (dataDirEnvVarForPackage distDirLayout))
                            . InstallPlan.toList
 
 -- | Construct an environment variable that points
@@ -393,19 +394,27 @@ dataDirsEnvironmentForPlan = catMaybes
 --   for inplace packages.
 -- * 'Nothing' for packages installed in the store (the path was
 --   already included in the package at install/build time).
--- * The other cases are not handled yet. See below.
-dataDirEnvVarForPackage :: ElaboratedConfiguredPackage
+dataDirEnvVarForPackage :: DistDirLayout
+                        -> ElaboratedConfiguredPackage
                         -> Maybe (String, Maybe FilePath)
-dataDirEnvVarForPackage pkg =
-  case (elabBuildStyle pkg, elabPkgSourceLocation pkg)
-  of (BuildAndInstall, _) -> Nothing
-     (BuildInplaceOnly, LocalUnpackedPackage path) -> Just
-       (pkgPathEnvVar (elabPkgDescription pkg) "datadir",
-        Just $ path </> dataDir (elabPkgDescription pkg))
-     -- TODO: handle the other cases for PackageLocation.
-     -- We will only need this when we add support for
-     -- remote/local tarballs.
-     (BuildInplaceOnly, _) -> Nothing
+dataDirEnvVarForPackage distDirLayout pkg =
+  case elabBuildStyle pkg
+  of BuildAndInstall -> Nothing
+     BuildInplaceOnly -> Just
+       ( pkgPathEnvVar (elabPkgDescription pkg) "datadir"
+       , Just $ srcPath (elabPkgSourceLocation pkg)
+            </> dataDir (elabPkgDescription pkg))
+  where
+    srcPath (LocalUnpackedPackage path) = path
+    srcPath (LocalTarballPackage _path) = unpackedPath
+    srcPath (RemoteTarballPackage _uri _localTar) = unpackedPath
+    srcPath (RepoTarballPackage _repo _packageId _localTar) = unpackedPath
+    srcPath (RemoteSourceRepoPackage _sourceRepo (Just localCheckout)) = localCheckout
+    -- TODO: see https://github.com/haskell/cabal/wiki/Potential-Refactors#unresolvedpkgloc
+    srcPath (RemoteSourceRepoPackage _sourceRepo Nothing) = error
+      "calling dataDirEnvVarForPackage on a not-downloaded repo is an error"
+    unpackedPath =
+      distUnpackedSrcDirectory distDirLayout $ elabPkgSourceId pkg
 
 instance Package ElaboratedConfiguredPackage where
   packageId = elabPkgSourceId

